buildkite/src/Lib/SelectFiles.dhall

```diff
@@ -0,0 +1,86 @@
+-- A DSL for representing file globbing; compiles into egrep-compatible regexp
+let Prelude = ../External/Prelude.dhall
+let P = Prelude
+
+let PathPattern = < Lit : Text | Any > -- Any == *
+let FilePattern = {
+  Type = { dir: Optional (List PathPattern), exts: Optional (List Text) },
+  default = { dir = None (List PathPattern), exts = None (List Text) }
+}
+
+let everything : FilePattern.Type = FilePattern.default
+
+let contains : Text -> FilePattern.Type =
+  \(s : Text) ->
+  { dir = Some [PathPattern.Lit s],
+    exts = None (List Text)
+  }
+
+let exactly : Text -> Text -> FilePattern.Type =
+  \(dir : Text) -> \(ext : Text) ->
+    { dir = Some [PathPattern.Lit dir],
+      exts = Some [ext]
+    }
+
+-- compile things
+let any : List Text -> Text =
+  P.Text.concatSep "|"
+
+let compileExt : Text -> Text =
+  \(ext : Text) -> "\\." ++ ext ++ "\\\$"
+
+let compileExts : List Text -> Text =
+  \(exts : List Text) ->
+    let exts_ = P.List.map Text Text compileExt exts in
+    "(" ++ (any exts_) ++ ")"
+
+let compilePath : PathPattern -> Text =
+  \(path : PathPattern) ->
+    merge {
+      Lit = \(text : Text) -> text,
+      Any = ".*"
+    } path
+
+let compilePaths : List PathPattern -> Text =
+  \(paths : List PathPattern) ->
+    let paths_ = P.List.map PathPattern Text compilePath paths in
```

Comment:
You may actually want to split the paths on "/" and filter out empty entries before recombining the pieces back to into the final string.

Reply:
Unfortunately, dhall cannot split strings only join them. We could require the components be passed in a list, and I thought about this a bit, but I think it'd be too annoying for the benefit.